darshan-util/darshan-mpiio-logutils.c

```diff
@@ -320,16 +320,45 @@ static void darshan_log_agg_mpiio_files(void *rec, void *agg_rec, int init_flag)
     int64_t tmp_val[4];
     int64_t tmp_cnt[4];
     int tmp_ndx;
-    double old_M;
     double mpi_time = mpi_rec->fcounters[MPIIO_F_READ_TIME] +
         mpi_rec->fcounters[MPIIO_F_WRITE_TIME] +
         mpi_rec->fcounters[MPIIO_F_META_TIME];
     double mpi_bytes = (double)mpi_rec->counters[MPIIO_BYTES_READ] +
         mpi_rec->counters[MPIIO_BYTES_WRITTEN];
+    int shared_file_flag = 0;
+#if 0
+    /* NOTE: the code commented out in this function is used for variance
+     * calculation.  This metric is most helpful for shared file records,
+     * but this function has now been generalized to allow for aggregation
+     * of arbitrary records in a log.
+     *
+     * This functionality could be reinstated as an optional feature.
+     * Ideally in that case the caller would also provide a buffer for
+     * stateful calculations; the current logic assumes that it is safe to
+     * use additional bytes off of the end of the rec argument buffer.
+     */
     struct var_t *var_time_p = (struct var_t *)
         ((char *)rec + sizeof(struct darshan_mpiio_file));
     struct var_t *var_bytes_p = (struct var_t *)
         ((char *)var_time_p + sizeof(struct var_t));
+    double old_M;
+#endif
+    /* if this is our first record, store base id and rank */
+    if(init_flag)
+    {
+        agg_mpi_rec->base_rec.rank = mpi_rec->base_rec.rank;
+        agg_mpi_rec->base_rec.id = mpi_rec->base_rec.id;
+    }
+
+    /* so far do all of the records reference the same file? */
+    if(agg_mpi_rec->base_rec.id == mpi_rec->base_rec.id)
+        shared_file_flag = 1;
+    else
+        agg_mpi_rec->base_rec.id = 0;
+
+    /* so far do all of the records reference the same rank? */
+    if(agg_mpi_rec->base_rec.rank != mpi_rec->base_rec.rank)
+        agg_mpi_rec->base_rec.rank = -1;
```

Comment:
Probably worth pointing out that if the caller passes in records representing partial (i.e. not all ranks) access to a shared file, it's going to be assigned a `-1` rank, which has always been used to represent globally shared ranks in Darshan.

Reply:
Yeah, good call; that's a subtle issue.  I just added a big comment block to the agg_record fn ptr as a warning.